### ISSUE TITLE
Thread-local shared data

### ISSUE THREAD

**Issue by rinon** [immunant/IA2-Phase2#249]
I don't think we currently support the `IA2_SHARED` annotation on thread-local data. We just need some way to make sure this data is allocated to compartment 0, right?

We currently need thread-local shared data for at least signal alternate stacks, although for that we need to be careful that the _default_ pkru value can read/write the stack, since the kernel resets the PKRU to a default value before invoking the signal handler.

**Comment by fw-immunant** [comment]
> we need to be careful that the _default_ pkru value can read/write the stack, since the kernel resets the PKRU to a default value before invoking the signal handler.

I.e., 0x55555554?

**Comment by rinon** [comment]
Yes, `0x55555554` although I believe there is some kernel support for changing the default? https://elixir.bootlin.com/linux/latest/source/arch/x86/mm/pkeys.c#L168

**Comment by rinon** [comment]
init_pkru requires kernel 4.9 fwiw

**Comment by fw-immunant** [comment]
We discussed this a fair bit in a call. Takeaways:
- On the thread-local shared data side:
  - `ia2_stackptr_0` is basically our prototype for shared TLS data. Presently, we `dlsym()` it and avoid protecting its page in `protect_tls_data`. This [gives us a dependency on calling `dlsym()` in protect_tls_data`](https://github.com/immunant/IA2-Phase2/pull/229/files#r1273823909), which we don't want. It also doesn't scale to other shared thread-locals.
  - We can't simply put shared TLS data in a different ELF section, because we really need glibc to do its setup magic which it only does on the TLS sections it knows about.
  - I considered some kind of macro scheme using asm symbol names (`type foo __asm__("ia2_shared_tls_foo");`) so that we could at least detect shared TLS data by name by iterating+inspecting symbols. But I think this would break cross-library references, and iterating+inspecting symbols is likely to be very slow, and probably still has a libdl dependency.
  - If we can opt into the general dynamic TLS model (see tls.pdf) per-variable, that may offer a solution. This model involves calling `__tls_get_addr`, which is "basically" implemented as `__tls_get_addr(module, offset) = dtv[thread_id][module] + offset`. If we can guarantee that no non-shared TLS variables use the general dynamic TLS model, and also opt shared TLS variables into it, we can then overwrite `dtv[thread_id][module]` to point at unprotected, shared storage that we allocate at thread startup.

- On the signal-handling side:
  - `/sys/kernel/debug/x86/init_pkru` from v4.9 onward controls the systemwide pkru value used for signal handling and new processes. It's system-wide, so multiple processes using IA2 must not require different pkru values for startup/signal-handling.
  - We can't allow privileged compartments to handle signals without `sigaltstack`. This is because their registers will be pushed onto the current stack, which may be another compartment's. Other threads in that compartment can then alter that memory, modifying register state that we trust (for example, the value in `rax` before a `wrpkru`.
  - With `sigaltstack`, any compartment that gets a signal will read its saved register state off of the signal stack at return. This means that whatever compartments can modify the contents of the signal stack can also manipulate other compartments at-will. As such, we have to designate a "trusted" compartment that can interfere with other compartments if the process needs to have a signal handler.

From this discussion, it seems like signal handling probably doesn't actually need a shared thread-local, so perhaps we can punt on this for now if we can solve the `dlsym()`-in-`protect_tls_pages` problem.
